### PR TITLE
compiletest: Add testing for lint machinery at runtime

### PR DESCRIPTION
**Pull request by clubby789**
While implementing https://github.com/rust-lang/rust/pull/157707, I noticed that we don't have 'runtime' tests for the diagnostic machinery (i.e. we test the derive macros themselves, but not their output when emitted). This introduces a new way to test this machinery.

- Add `//@ aux-lint: file.rs`, which builds a lint library in `auxiliary/file.rs`
- Lint libraries link with `rustc_private`, and export a `extern "C" fn register_lints(...)`
- A new `compiletest-lint-driver` wraps rustc and installs the provided lints when building the test

Included is a simple demo lint for example

_AI Disclosure: I used an LLM to plan out the broad structure of the machinery and identify issues; all code in this PR is written by hand_

### CODE REVIEW

File: src/bootstrap/src/core/build_steps/test.rs [hunk 1]
```diff
@@ -2096,6 +2102,17 @@ NOTE: if you're sure you want to do this, please open an issue as to why. In the
         cmd.arg("--compile-lib-path").arg(builder.rustc_libdir(test_compiler));
         cmd.arg("--run-lib-path").arg(builder.sysroot_target_libdir(test_compiler, target));
         cmd.arg("--rustc-path").arg(builder.rustc(test_compiler));
+        if suite == "ui-fulldeps" {
+            let lint_driver: PathBuf = builder
+                .ensure(tool::CompiletestLintDriver {
+                    compiler: test_compiler,
+                    target: test_compiler.host,
+                })
+                .tool_path;
```
Comment (clubby789) [inline src/bootstrap/src/core/build_steps/test.rs:2111]: This does add some tool+std builds unconditionally to any `ui-fulldeps` invoke; this should hopefully be relatively low impact in practice

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust/157721/485ec3fbcc12fa14ef6596dabb125ad710499c9e..9c087122d0a7d718c6e5e45e735beb9f013feee0)*

**Comment by clubby789** [comment]
Just because I'm a little unsure on the correctness of some of the logic:
@bors try jobs=x86_64-gnu-stable,dist-x86_64-msvc

**Comment by Mark-Simulacrum** [comment]
> we test the derive macros themselves, but not their output when emitted

I'm a bit confused (probably just missing something...)... but aren't those macros tested by every diagnostics-related UI test in the existing suite? It's an indirect test but it should still catch bugs in practice, right?

If we wanted to test the raw output, I'd expect that to be done with something more like proc-macro2 + insta to snapshot test what is produced by the macro?

I think in general it would make sense to make a compiler MCP for this since it'll at minimum make it a bit more painful to edit the lint infrastructure.

**Comment by clubby789** [comment]
'Output' as in the runtime diagnostics emission. The machinery in general is tested, but when introducing new features to the system (i.e. lint MSRV) it would be nice to test the edge cases of behaviour.

An MCP probably does make sense here though.

**Comment by clubby789** [comment]
MCP: https://github.com/rust-lang/compiler-team/issues/1004
